File: app/src/main/java/com/nirbhor/app/domain/StockCalculator.kt
package com.nirbhor.app.domain

/** Shared stock math so add, cabinet, refill, and alerts cannot disagree. */
object StockCalculator {
    fun estimatedDays(
        stockCount: Int,
        dosePerIntake: Float,
        timesPerScheduledDay: Int,
        frequency: Frequency,
        weekdaysMask: Int,
        paused: Boolean = false,
    ): Int? {
        if (paused || stockCount < 0 || dosePerIntake <= 0f || timesPerScheduledDay <= 0) return null
        val scheduledDaysPerWeek = when (frequency) {
            Frequency.DAILY -> 7f
            Frequency.ALTERNATE -> 3.5f
            Frequency.WEEKDAYS, Frequency.WEEKLY -> Integer.bitCount(weekdaysMask and 0x7F).toFloat()
        }
        if (scheduledDaysPerWeek <= 0f) return null
        val averagePerDay = dosePerIntake * timesPerScheduledDay * scheduledDaysPerWeek / 7f
        return (stockCount.coerceAtLeast(0) / averagePerDay).toInt()
    }

    fun estimatedDays(medicine: Medicine): Int? = estimatedDays(
        stockCount = medicine.stockCount,
        dosePerIntake = medicine.dosePerIntake,
        timesPerScheduledDay = medicine.timeTokens.size,
        frequency = medicine.frequency,
        weekdaysMask = medicine.weekdaysMask,
        paused = medicine.paused,
    )
}

File: lib/domain/stock_calculator.dart
import 'dart:math' as math;

import 'models.dart';

/// Shared stock math so add, cabinet, refill, and alerts cannot disagree.
class StockCalculator {
  StockCalculator._();

  /// Units removed from stock by one intake. `stockCount` is an integer column shared with the
  /// Kotlin schema, so a half tablet costs a whole unit; the days estimate below uses the same
  /// number so the cabinet never promises more days than the count can actually deliver.
  static int unitsPerIntake(double dosePerIntake) => math.max(1, dosePerIntake.ceil());

  static int? estimatedDaysFrom({
    required int stockCount,
    required double dosePerIntake,
    required int timesPerScheduledDay,
    required Frequency frequency,
    required int weekdaysMask,
    bool paused = false,
  }) {
    if (paused || stockCount < 0 || dosePerIntake <= 0 || timesPerScheduledDay <= 0) return null;
    final double scheduledDaysPerWeek = switch (frequency) {
      Frequency.daily => 7,
      Frequency.alternate => 3.5,
      Frequency.weekdays || Frequency.weekly => _bitCount(weekdaysMask & 0x7F).toDouble(),
    };
    if (scheduledDaysPerWeek <= 0) return null;
    final averagePerDay =
        unitsPerIntake(dosePerIntake) * timesPerScheduledDay * scheduledDaysPerWeek / 7;
    return ((stockCount < 0 ? 0 : stockCount) / averagePerDay).floor();
  }

  static int? estimatedDays(Medicine medicine) => estimatedDaysFrom(
        stockCount: medicine.stockCount,
        dosePerIntake: medicine.dosePerIntake,
        timesPerScheduledDay: medicine.timeTokens.length,
        frequency: medicine.frequency,
        weekdaysMask: medicine.weekdaysMask,
        paused: medicine.paused,
      );

  static int _bitCount(int v) {
    var n = v;
    var c = 0;
    while (n != 0) {
      c += n & 1;
      n >>= 1;
    }
    return c;
  }
}
